File: fe_camphub/src/libs/core/mapping/booking.mapper.ts
// libs/mappers/booking.mapper.ts
import { createMapper, createMap, forMember, mapFrom } from "@automapper/core";
import { pojos, PojosMetadataMap } from "@automapper/pojos";

import { Booking } from "../types";
import { BookingResponse } from "../dto/response";
import {
    BookingCreationRequest,
    BookingItemRequest,
    LesseeReturnRequest,
} from "../dto/request";

export const bookingMapper = createMapper({
    strategyInitializer: pojos(),
});

PojosMetadataMap.create<Booking>("Booking", {
    id: String,
    lesseeId: String,
    lessorId: String,
    itemId: String,
    itemName: String,
    lessorName: String,
    lesseeName: String,
    quantity: Number,
    pricePerDay: Number,
    depositAmount: Number,
    totalAmount: Number,
    startDate: String,
    endDate: String,
    note: String,
    status: String,
    createdAt: String,
    hasReviewed: Boolean,
});

PojosMetadataMap.create<BookingResponse>("BookingResponse", {
    id: String,
    lesseeId: String,
    lessorId: String,
    itemId: String,
    itemName: String,
    lessorName: String,
    lesseeName: String,
    quantity: Number,
    pricePerDay: Number,
    depositAmount: Number,
    totalAmount: Number,
    startDate: String,
    endDate: String,
    note: String,
    status: String,
    createdAt: String,
    hasReviewed: Boolean,
});

PojosMetadataMap.create<BookingCreationRequest>("BookingCreationRequest", {
    items: Array,
});

PojosMetadataMap.create<BookingItemRequest>("BookingItemRequest", {
    cartItemId: String,
    startDate: String,
    endDate: String,
    quantity: Number,
    pricePerDay: Number,
    depositAmount: Number,
    note: String,
});

PojosMetadataMap.create<LesseeReturnRequest>("LesseeReturnRequest", {
    bookingId: String,
    note: String,
    mediaUrls: Array,
});

// Mapping chính
createMap<BookingResponse, Booking>(bookingMapper, "BookingResponse", "Booking");

createMap<Booking, BookingCreationRequest>(
    bookingMapper,
    "Booking",
    "BookingCreationRequest",
    forMember(
        (d) => d.items,
        mapFrom((s) => [
            {
                cartItemId: s.itemId,
                startDate: s.startDate,
                endDate: s.endDate,
                quantity: s.quantity,
                pricePerDay: s.pricePerDay,
                depositAmount: s.depositAmount,
                note: s.note ?? "",
            } as BookingItemRequest,
        ])
    )
);

createMap<Booking, LesseeReturnRequest>(
    bookingMapper,
    "Booking",
    "LesseeReturnRequest",
    forMember((d) => d.bookingId, mapFrom((s) => s.id)),
    forMember((d) => d.note, mapFrom((s) => s.note ?? "")),
    forMember((d) => d.mediaUrls, mapFrom(() => []))
);

createMap<Booking, BookingResponse>(bookingMapper, "Booking", "BookingResponse");

// Export mapper helper
export const bookingMap = {
    fromResponse: (dto: BookingResponse): Booking =>
        bookingMapper.map<BookingResponse, Booking>(dto, "BookingResponse", "Booking"),

    toCreationRequest: (model: Booking): BookingCreationRequest =>
        bookingMapper.map<Booking, BookingCreationRequest>(model, "Booking", "BookingCreationRequest"),

    toLesseeReturn: (model: Booking): LesseeReturnRequest =>
        bookingMapper.map<Booking, LesseeReturnRequest>(model, "Booking", "LesseeReturnRequest"),

    toResponse: (model: Booking): BookingResponse =>
        bookingMapper.map<Booking, BookingResponse>(model, "Booking", "BookingResponse"),
};

File: be_camphub/src/main/java/org/camphub/be_camphub/mapper/BookingMapper.java
package org.camphub.be_camphub.mapper;

import org.camphub.be_camphub.dto.request.booking.BookingItemRequest;
import org.camphub.be_camphub.dto.response.booking.BookingResponse;
import org.camphub.be_camphub.entity.Booking;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring")
public interface BookingMapper {

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "lesseeId", ignore = true) // sẽ set trong service
    @Mapping(target = "lessorId", ignore = true) // lấy từ CartItem
    @Mapping(target = "itemId", ignore = true) // lấy từ CartItem
    @Mapping(target = "status", constant = "PENDING_CONFIRM")
    @Mapping(target = "createdAt", expression = "java(java.time.LocalDateTime.now())")
    @Mapping(target = "updatedAt", ignore = true)
    Booking creationRequestToEntity(BookingItemRequest request);

    @Mapping(target = "itemName", ignore = true)
    @Mapping(target = "lessorName", ignore = true)
    @Mapping(target = "lesseeName", ignore = true)
    @Mapping(target = "totalAmount", expression = "java(booking.getPricePerDay() * booking.getQuantity())")
    BookingResponse entityToResponse(Booking booking);
}
